[유안] API 테스트 / 문서자동화 미션 제출합니다.

KimSeongGyu1 opened the pull request:
잘 부탁드립니다

1 멤버 관리 / 즐겨찾기에 관한 코드에만 수정을 했습니다

2 학습을 위해 세션과 토큰을 동시에 사용하는 연습을 해보았습니다
    - Resolver에서 이 둘이 같은 값인지에 대한 검증을 해주었습니다

3 한 리소스에 대해서 한 개의 uri만을 가지게 하고 싶어서 "/me" 를 "/members/id"로 변경했습니다
    - uri의 id값과 세션, 토큰의 정보가 일치하는지 Resolver에서 검증해줍니다

toneyparky committed:
docs: 기능목록 작성

toneyparky committed:
practice: Basic Authentication 인터셉터 실습

toneyparky committed:
practice: Bearer Authentication 인터셉터 실습

toneyparky committed:
practice: Session Authentication 인터셉터 실습

toneyparky committed:
practice: 문서 자동화 실습

toneyparky committed:
chore: 실습에서 생성한 미션에 불필요한 파일 제거

WebMvcConfig: Basic 방식 제거
LoginMemberController: 사용자 정보 조회 Basic URI 제거
AuthAcceptanceTest: Basic 방식 테스트 제거

toneyparky committed:
feat: 클라이언트 회원가입 폼 검증

/snackbar: alert 대신 사용할 snackbar 생성
Join.js: 유효성 검사 메서드 생성

toneyparky committed:
feat: 회원가입 기능 클라이언트 연동

MemberRequest: 생성자 추가
index.js: member 관련 라우트 추가
join.js: 멤버 생성 API 요청

toneyparky committed:
refactor: admin 관련 페이지 라우팅 수정

PageController: "/"를 사용자 메인 페이지로 수정

toneyparky committed:
feat: 로그인 기능 클라이언트 연동

LoginMemberController: JWT 토큰과 Session을 함께 발행
index.js: find를 POST로 변경
Login.js: 로그인 fetch api 연동, 사용자 메인 페이지로 리다이렉트 추가
Join.js: 사용자 메인 페이지로 리다이렉트 추가

toneyparky committed:
feat: 내 정보 조회 기능 클라이언트 연결

LoginMemberController: 내 정보 조회 기능 라우팅 통합
LoginMemberMethodArgumentResolver: 세션과 JWT 회원정보 일치 여부 비교
SessionInterceptor, BearerAuthInterceptor: 회원정보 일치 여부 비교를 위한 요청 키 값 수정
Login.js: 로그인시 JWT 값을 localStorage에 저장
index.js: 헤더에 Authorization을 담아 요청
PageController: 내 정보 조회 라우팅

toneyparky committed:
feat: 내 정보 수정 기능 클라이언트 연결

index.js: 헤더에 Authorization을 담아 요청
mypage-edit.html: email 수정 못하도록 p 태그로 변경, data-id 속성 추가

toneyparky committed:
feat: 내 정보 삭제 기능 클라이언트 연결

index.js: 헤더에 Authorization을 담아 요청
MyPageEdit.js: 버튼에 삭제 이벤트 연결

toneyparky committed:
feat: 미 로그인시 내 정보 조회 접근 블락

ValidateLogin.js: 로그인 확인 로직 구현
index.js: 회원 정보 삭제 시 localStorage 삭제
myPage.html, myPage-edit.html: 로그인 validation js 삽입

toneyparky committed:
feat: 로그인 여부에 따른 메인 페이지 메뉴 랜더링

templates.js: 로그인 여부에 따른 동적 템플릿 구성
index.js: 로그인 여부에 따라 적절한 템플릿 호출

toneyparky committed:
feat: 로그아웃 구현

Index.js: 로그아웃 버튼 클릭시 localStorage 초기화 및 리로딩

toneyparky committed:
fix: 깨지는 테스트 수정

AuthAcceptanceTest: session과 bearer을 동시에 검증하도록 수정
PageAcceptanceTest: MediaType 수정

toneyparky committed:
feat: 회원 정보 조회 문서화

MemberControllerTest: readMember 테스트
MemberDocumentation, api-guide.adoc: 생성할 API 문서 설정

toneyparky committed:
feat: 회원 정보 수정 문서화

MemberControllerTest: updateMember 테스트
MemberDocumentation: updateMember 메서드 수정

toneyparky committed:
feat: 회원 탈퇴 문서화

MemberControllerTest: deleteMember 테스트
MemberDocumentation: deleteMember 메서드 구현

toneyparky committed:
feat: /members 경로에 대한 유효성 검사 적용

AOP를 활용해서 Interceptor에서 설정해준 attribute를 이용하여 유효성 검사
어노테이션을 활용해 미적용할 메서드 지정
테스트도 수정사항에 맞추어 변경

toneyparky committed:
feat: 커스텀 예외 생성 및 Advice를 이용한 처리

커스텀 예외를 사용하도록 코드 수정
PathService: NoPathExistsException를 던지는 예외처리 추가

toneyparky committed:
feat: 회원 Resource 접근시 validation 적용

관련 예외에 대한 문서화 진행

toneyparky committed:
refactor: 사용자 검증 방식 변경

AOP 사용하던 부분을 MethodArgumentResolver로 변경

toneyparky committed:
feat: 즐겨찾기 도메인 생성

Favorite: 즐겨찾기 도메인 클래스 생성
Favorites: 일급 컬렉션 생성
Member: Favorites 필드 추가

toneyparky committed:
feat: 즐겨찾기 서비스 구현

MemberService: 즐겨찾기 추가, 제거 메서드 구현
Favorite: 정적 생성자 메서드 추가
Favorites, Member: 사용하지 않는 메서드 제거

toneyparky committed:
docs: 즐겨찾기 기능에 대한 README.md 업데이트

toneyparky committed:
refactor: FavoriteService 분리

toneyparky committed:
feat: 즐겨찾기 추가 API 구현

FavoriteController: 즐겨찾기 추가 라우팅
AddFavoriteRequest, FavoriteResponse: dto 생성
WebMvcConfig: 즐겨찾기 uri를 인터셉터에 추가
FavoriteService: addFavorite시 추가된 즐겨찾기를 반환하도록 수정

toneyparky committed:
feat: 즐겨찾기 조회 API 구현

FavoriteController: 즐겨찾기 조회 라우팅
FavoritesResponse: dto 생성
FavoriteService: readFavorites 메서드 구현

toneyparky committed:
feat: 즐겨찾기 삭제 API 구현

FavoriteController: 즐겨찾기 삭제 라우팅
FavoriteService: removeFavorite 메서드 구현

toneyparky committed:
refactor: 즐겨찾기 조회 API 응답 수정

id로만 응답하던 내용을 역 이름까지 응답하도록 수정

toneyparky committed:
test: 즐겨찾기 인수 테스트

toneyparky committed:
feat: 즐겨찾기 추가 문서화

truncate.sql: 테이블 truncate시 아이디를 1로 초기화
FavoriteDocumentation: addFavorite메서드 구현

toneyparky committed:
feat: 즐겨찾기 조회 문서화

FavoriteDocumentation: readFavorite메서드 구현

toneyparky committed:
feat: 즐겨찾기 삭제 문서화

FavoriteDocumentation: deleteFavorite메서드 구현

toneyparky committed:
feat: 중복된 즐겨찾기 추가에 대한 방어기능 구현

DuplicatedFavoriteException: 커스텀 예외 생성
FavoriteService: 즐겨찾기 추가시 중복된 경우에 대해 예외처리
SubwayControllerAdvice: 예외 핸들링

toneyparky committed:
chore: 초기 데이터 추가

toneyparky committed:
fix: 슬라이드 깨짐 수정

toneyparky committed:
feat: 즐겨찾기 추가 프론트 연동

LoginMemberController: login시 location 속성으로 id를 전달

toneyparky committed:
feat: 즐겨찾기 조회 및 삭제 프론트 연동

toneyparky committed:
refactor: 컨벤션과 가독성을 향상시키기 위한 수정

RequestMapping으로 공통된 url 중복 제거
ResponseEntity에 제네릭 Type 명시

toneyparky committed:
refactor: 사용자 API 경로를 /mypage 에서 members/{id}로 수정

toneyparky committed:
refactor: 사용자 API 경로를 /login 에서 members/login으로 수정

패키지 변경

toneyparky committed:
refactor: 중복된 이메일로 회원가입 방지

KimSeongGyu1 committed:
refactor: uri path 검증 로직 Resolver로 이동

KimSeongGyu1 committed:
refactor: 인수테스트 보강

가독성위해 주석 추가
삭제에 대한 테스트 추

KimSeongGyu1 committed:
refactor: clean up code

dave915 commented on src/main/java/wooteco/subway/web/advice/SubwayControllerAdvice.java:
```diff
@@ -0,0 +1,45 @@
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<ExceptionResponse> handleUnexpectedException(Exception e) {
+        LOGGER.error(e);
+        return new ResponseEntity<>(ExceptionResponse.of("서버 오류가 발생했어요."), HttpStatus.INTERNAL_SERVER_ERROR);
```
👍

dave915 commented on src/main/java/wooteco/subway/web/advice/SubwayControllerAdvice.java:
```diff
@@ -0,0 +1,45 @@
+
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<ExceptionResponse> handleUnexpectedException(Exception e) {
+        LOGGER.error(e);
```
log를 남길때 해당하는 log를 찾기 쉽도록
logger.error("error message >>> {}", error.getMessage(), e);
이런 식으로 앞에 error message >>> 같은 문구를 넣어주거나 에러에 관한 더 자세한 정보를 함께 남길 수 있어요 :)

아래 글을 참고 해 보세요 :)
https://blog.leocat.kr/notes/2018/12/30/logger-last-throwable-param-on-logging-formatted-message

KimSeongGyu1 replied:
log를 잘보이도록 하는 김에 더 잘보이도록 줄바꿈과 특수문자를 섞어서

LOGGER.error("\n@@@@@@@@@@ error message @@@@@@@@@@\n >>> {}", e.getMessage(), e);

이렇게 바꿔 봤어요.
저는 이게 더 잘 보여서서 좋은데 이런식으로 해도 문제가 없을까요?

dave915 replied:
문제는 없지만 보통 로그를 서버에 파일로 남기고, 이는 서버의 리소스를 사용하는 것이기 때문에, 왠만하면 필요한 문자만 사용하는 것이 좋을 것 같아요 :) 
생각보다 서버에서 disk full이 자주 일어나는 것 같아요.

dave915 commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthInterceptor.java:
```diff
@@ -22,27 +22,24 @@ public BearerAuthInterceptor(AuthorizationExtractor authExtractor, JwtTokenProvi
-                           Object handler,
-                           ModelAndView modelAndView) throws Exception {
+        if (StringUtils.isEmpty(token)) {
+            throw new InvalidAuthenticationException("토큰이 없어요");
```
토큰이 없을때를 테스트 하는 코드가 없는것 같아요 :)

KimSeongGyu1 replied:
컨트롤러에서만 테스트를 하고 인지 테스트를 하지 않았네요ㅜ
추가했습니다

dave915 commented on src/test/java/wooteco/subway/web/member/MemberControllerTest.java:
```diff
@@ -1,5 +1,227 @@
+        Member member = new Member(1L, TEST_USER_EMAIL, TEST_USER_NAME, TEST_USER_PASSWORD);
+        given(memberService.createMember(any())).willReturn(member);
+
+        String inputJson = "{\"email\":\"" + TEST_USER_EMAIL + "\"," +
+                "\"name\":\"" + TEST_USER_NAME + "\"," +
+                "\"password\":\"" + TEST_USER_PASSWORD + "\"}";
```
이 부분이 중복되어 보여요!

KimSeongGyu1 replied:
String.format()으로 처리했어요

dave915 commented on src/test/java/wooteco/subway/web/member/MemberControllerTest.java:
```diff
@@ -1,5 +1,227 @@
+        MockHttpSession session = new MockHttpSession();
+        session.setAttribute("loginMemberEmail", TEST_USER_EMAIL);
+
+        this.mockMvc.perform(get("/members?email=" + TEST_USER_EMAIL)
+                .session(session)
+                .header("authorization", "Bearer " + TEST_USER_TOKEN)
+                .accept(MediaType.APPLICATION_JSON)
+                .contentType(MediaType.APPLICATION_JSON))
```
이부분도 중복 제거 가능하지 않을까요?

KimSeongGyu1 replied:
상속 형태로 FavoriteControllerTest까지 중복제거 해봤어요

dave915 commented on src/main/java/wooteco/subway/domain/member/Member.java:
```diff
@@ -2,13 +2,18 @@
     private String name;
     private String password;
+    @Embedded.Empty
+    private Favorites favorites = Favorites.empty();
```
Member가 favorites를 갖고있는 것이 맞는가에 대해 의문이 드는데, 유안은 어떻게 생각하시나요?

KimSeongGyu1 replied:
1)
Member가 favorites를 가지지 않는다면 따로 id값을 가지는 엔티티가 될텐데,
이 때 가장 큰 차이점은 Member가 삭제될 때 Favorites이 같이 삭제가 되는가 라고 생각해요.

계속 누적되는 즐겨찾기 데이터를 가지고 유의미한 정보를 가공하고 싶다면 Favorites 데이터가 삭제되지 않도록 막아야 한다고 생각해요.
사용자가 탈퇴했을 때 개인 정보를 괜히 계속 들고 있을 필요가 있는가라고 생각했을 때, 지금은 필요 없다고 생각했어요.

2)
현재 Spring data jdbc를 어설프게 사용하고 있어서 생기는 문제가 있어요
단순히 Member의 password 하나만 수정해도 Member의 모든 값에 대한 쿼리가 날아가기 때문에 Member가 favorites를 가지면 성능 이슈가 있어요
커스텀 쿼리를 만들어서 해결할 수 있다고 들었는데 이 부분은 한 번 공부해서 적용해볼게요

dave915 replied:
Member가 삭제될 때 Favorites이 같이 삭제하는 것은 Member가 꼭 Favorites를 필드로 갖고있지 않아도 로직적으로 처리하거나, DB기능으로 처리 가능해 보여요. 

Member를 조회할때 Favorites를 함께 조회하는가에 대해 생각해보면 어떨까요? 
그리고, Favorites같은 테이블이 많아졌을때 지금처럼 Member가 필드로 들고있으면 어떤 문제가 있을지 생각해 보면 좋을 것 같아요 :)

dave915 commented on src/main/java/wooteco/subway/domain/member/Member.java:
```diff
@@ -54,4 +59,36 @@ public void update(String name, String password) {
+
+    public void validateId(Long id) {
+        if (!this.id.equals(id)) {
+            throw new InvalidAuthenticationException("니 아이디 아님");
```
`니 아이디 아님` 인지 테스트코드가 없는 것 같아요 😁
전체적으로 Exception에관한 테스트가 없어보이는 것 같아요

KimSeongGyu1 replied:
전반적으로 Exception에 관한 테스트를 추가했어요
리뷰를 받고 우연히 바로 책에서 Exception 테스트의 중요성에 대해 읽었는데, 중요성이 더 와닿아서 좋았어요

dave915 commented on src/main/java/wooteco/subway/config/WebMvcConfig.java:
```diff
@@ -4,34 +4,29 @@
-        registry.addInterceptor(basicAuthInterceptor).addPathPatterns("/me/basic");
-        registry.addInterceptor(sessionInterceptor).addPathPatterns("/me/session");
-        registry.addInterceptor(bearerAuthInterceptor).addPathPatterns("/me/bearer");
+        registry.addInterceptor(sessionInterceptor).addPathPatterns("/members", "/members/**");
+        registry.addInterceptor(bearerAuthInterceptor).addPathPatterns("/members", "/members/**");
```
지금은 members에 관해서만 인터셉터가 걸렸지만, 나중에 api가 많아졌을땐 전체 Path를 걸고 인증처리 하지 않을 api만 exclude 해주는 것이 인증 누락될 확률이 낮을 것 같아요 :)

KimSeongGyu1 replied:
실수가 일어났을 때 인증처리를 할 곳에 안하는 상황보다는, 인증처리를 안 할 곳에 하는 것이 더 안전해 보이네요 :)

그런데 궁금한 점이 있어요
api가 많아졌을 때 path별로 다른 인증방식을 사용하는 경우가 있을까요?
그런 상황에서는 path별로 인증을 걸어주는 방향으로 하고 싶을 것 같아요

dave915 replied:
>api가 많아졌을 때 path별로 다른 인증방식을 사용하는 경우

이런 경우는 거의 없는 것 같아요. 있더라도 기본적인 인증방식을 먼저 전체에 걸고 필요에 따라 몇몇 Path에 다른 인증방식을 걸지 않을까싶어요. 

전체 path에 걸고 exclude하는 방식은 제안일 뿐이라서 필요에 따라 맞게 사용하시면 될 것 같아요 :)

dave915 commented on src/main/java/wooteco/subway/web/member/MemberController.java:
```diff
@@ -4,43 +4,70 @@
 
-    @PutMapping("/members/{id}")
-    public ResponseEntity<MemberResponse> updateMember(@PathVariable Long id, @RequestBody UpdateMemberRequest param) {
+    @GetMapping("/{id}")
```
코멘트 남겨주신 `/me -> /members/{id}` 부분이 이 부분일까요? 
제 생각엔 다른 사람의 정보를 조회 할 수 있도록 Api가 열린 것 같아서 보안 취약점이 될 수도 있을 것 같다는 생각이 드네요 ㅎㅎ 
물론 인증 처리로 토큰이나 세션과 해당 {id}가 같은가? 에 대한 체크를 해줘도 될 것 같긴 하지만 불필요하게 한번 더 하는것이 아닌가? 라는 생각이 들어요

KimSeongGyu1 replied:
네 지금은 토큰, 세션, id가 전부 일치하는지 체크하고 있어요
말씀하신대로 불필요할 수도 있다는 생각이 들어요

그런데 REST에 대해 신경쓰기 시작하면서 "/me"는 리소스를 제대로 표현하지 못한다고 생각해서 "members/id"로 했어요
"/me"는 자주 쓰이는 uri 패턴이라고 듣기는 했는데 이게 rest하지는 않지만 효율적이라서 사용하는 것인지, rest한지 잘 모르겠어요.
이 부분은 어떻게 검색해야 답을 얻을 수 있는지도 잘 모르겠어요ㅜ

dave915 replied:
몇가지 찾아봤는데 괜찮다고 나오네요! 


https://stackoverflow.com/questions/36520372/designing-uri-for-current-logged-in-user-in-rest-applications
https://stackoverflow.com/questions/35719797/is-using-magic-me-self-resource-identifiers-going-against-rest-principles

dave915 reviewed (changes requested):
유안 안녕하세요! 리뷰가 늦었네요 ㅠㅠ
몇가지 피드백 남겼으니 확인 후 반영해보세요 :)

KimSeongGyu1 committed:
refactor: 가독성 위한 log 출력 format 변환

KimSeongGyu1 committed:
refactor: 컨트롤러 테스트 중복 제거

- MemberController, FavoriteController에 대해 개선
공통되는 MockMvc에 대한 부분 묶어서 상속하도록 변경

- 그 외 String format 이용 중복 개선

KimSeongGyu1 committed:
refactor: 컨트롤러 인가 테스트 위치 변경

MemberControllerTest안에 있던 인가에 대한 테스트 상위 클래스로 위치 변경

KimSeongGyu1 committed:
refactor: 유저 인지 테스트 추가

테스트를 세분화 한 뒤 DynamicTest를 이용해 각자 검증
인가에 대해 부족한 테스트를 DynamicTest에 추가

KimSeongGyu1 committed:
refactor: 즐겨찾기 인지 테스트 리펙토링

DynamicTest 이용

KimSeongGyu1 committed:
refactor: 멤버 인지 테스트에서 인가 인지 테스트 분리

인가에 대한 부분은 따로 검증
인가 인지 테스트에 자기 아이디가 아닐 때 검증 추가

KimSeongGyu1 committed:
refactor: Exception 세분화 및 패키지 분리

Authentication 관련 예외 세분화

KimSeongGyu1 committed:
refactor: 404에 대한 인지테스트 추가

NoMemberExistException
NoFavoriteExistException

KimSeongGyu1 committed:
refactor: 400에 대한 인지테스트 추가

DuplicatedEmailException
DuplicatedFavoriteException

dave915 reviewed (changes requested):
유안! 테스트를 꼼꼼히 추가해주신 듯 하네요 👍 
기존 남겼던 코멘트만 읽어보셔도 될 것 같아요. 
추가적으로 질문이 있으시면 편하게 코멘트나 DM주세요 :)

KimSeongGyu1 committed:
refactor: JPA로 변환 (station)

KimSeongGyu1 committed:
refactor: JPA로 변환 (line, linestation)

lineStation은 line과 함께 eager로 로딩되는 벨류로 취급함

KimSeongGyu1 committed:
refactor: JPA로 변환 (member, favorite)

Favorite는 Member(member_id), Station(source_id, target_id)을 ManyToOne으로 참조

KimSeongGyu1 committed:
refactor: JPA로 변환 관련 테스트 수정

인수테스트 제외 domain, service, controller까지 테스트 통과 확인

KimSeongGyu1 committed:
refactor: Favorites 인수테스트 통과

FavoriteService 적절하게 수정
통과하지 못하는 인수테스트 일단 주석처리

KimSeongGyu1 committed:
refactor: 모든 인수테스트 통과하도록 수정

1. @Transactional 문제(jdbc 이용할 때 작성했던 불필요한 save문 제거도 진행)
2. Long과 long으로 인한 문제: (더티 체킹 후 delete 쿼리가 의도하지 않은 값에 대해서만 비교하는 문제)

KimSeongGyu1 committed:
refactor: 사용하지 않는 코드 제거

KimSeongGyu1 committed:
refactor: n + 1 문제 수정

dave915 commented on src/main/java/wooteco/subway/domain/line/Line.java:
```diff
@@ -21,13 +28,14 @@
     private LocalDateTime createdAt;
     @LastModifiedDate
     private LocalDateTime updatedAt;
```
JPA의 Audit기능에 대해 한번 읽어보셔도 좋을 것 같아요 :) 
https://tramyu.github.io/java/spring/jpa-auditing/

dave915 reviewed (approved):
유안 안녕하세요! 
미션 하느라 고생 많으셨습니다 👍
여기서 머지 해도 될 것 같아요. 코멘트 하나 남겼으니 확인해보세요~!
